### ISSUE TITLE
Improve documentation for with examples for JSON type props

### ISSUE THREAD

**Issue by icex** [mikro-orm/mikro-orm#1395]
**Describe the bug**
I'm a noob with Mikro ORM so don't judge too harshly if this may seem stupid. I have an issue where changes in the JSON prop are not registered. The Json is an array of objects in my case.

https://github.com/mikro-orm/mikro-orm/blob/c87d8001f17f72296dcf203afe212572edd25d48/packages/core/src/utils/Utils.ts#L31
```
    if (a.toString !== Object.prototype.toString) {
        return a.toString() === b.toString(); <-- `if a.toString()` and `b.toString()` return  `[object Object]` then this is always TRUE
    }
```

The only way I was able to get past this was to add a custom toString function on that object(not sure it's the best way):
```
    public toString(): string {
        return JSON.stringify(this);
    }
```

**To Reproduce**
You can test this behavior with a parameter that is populated with an array of objects.

**Expected behavior**
Compared json objects should be better handled.

**Versions**

| Dependency | Version |
| - | - |
| node | v14.15.4 |
| typescript | 4.1.3 |
| mikro-orm | 4.4.2 |
| your-driver | mongodb |

**Comment by B4nan** [comment]
Well, that method is called `compareObjects`, and that is what it should be used for - comparing objects, not arrays.

Please provide actual reproduction. There are tests for this.

**Comment by icex** [comment]
Thank you kindly for your reply. I was using type: "json" for that property. If I changed it to "array" it worked fine. Not sure if this makes sense, but shouldn't it have worked with "json" as well? When I think json I think any kind of data either arrays, strings, objects, bools etc..

**Comment by B4nan** [comment]
It should, but the issue is not the code you mentioned, the issue is that the method should not be used to compare arrays. 

I think I see where this is coming from, should be a one line fix. But to verify it works, we need a reproduction - that is what I am asking for. Already tried myself but it works fine for me.

**Comment by B4nan** [comment]
Still waiting for the repro, if you don't plan to provide it, we can close the issue as I have no interest in shipping blind fixes to things I can't reproduce :] 

Should be enough to provide the actual value and how you changed it when it was not recognized as a change.

**Comment by icex** [comment]
Really sorry for the late reply.
`
import {Embeddable, Entity, PrimaryKey, Property, SerializedPrimaryKey} from "@mikro-orm/core";
import {ObjectID} from "mongodb";

@Entity()
export class TestTemplate {

    @PrimaryKey({type: String})
    public _id: ObjectID;

    @SerializedPrimaryKey()
    public readonly id!: string;

    @Property()
    public name: string;

    @Property({type: "json", nullable: true}) // <---- here is what caused this issue -> replacing with json with array fixed it as you suggested.
    public messages?: EmailMessageTest[];
}

@Embeddable()
export class EmailMessageTest {
    @Property({nullable: true})
    public html?: string;

    @Property()
    public language: string = "en";
}

`

`
                const data = {
                    name: "test",
                    messages: [
                        {html: "aaa", language: "en"},
                        {html: "bbb", language: "fr"},
                    ],
                };
                const item = em.create(TestTemplate, data);
                await em.persistAndFlush(item);

                console.log(item);

                const data2 = {
                    name: "test new name",
                    messages: [
                        {html: "xxx", language: "en"},
                        {html: "yyy", language: "fr"},
                    ],
                };
                em.assign(item, data2);
                await em.persistAndFlush(item);

                console.log(item);
`

Actual data in my mongo instance:
`
{
    "name": "test new name",
    "messages": [{
        "html": "aaa",
        "language": "en"
    }, {
        "html": "bbb",
        "language": "fr"
    }]
}
`

As you can see this doesn't fail, it updates the `name` field, but the messages are silently disregarded.


Logs:
`
[07:44:17.700Z] INFO  EMAILTEMPLATES : [query] db.getCollection('test-template').insertOne({ name: 'test', messages: [ { html: 'aaa', language: 'en' }, { html: 'bbb', language: 'fr' } ] }, { session: undefined }); [took 3 ms]
TestTemplate {
  name: 'test',
  messages: [ { html: 'aaa', language: 'en' }, { html: 'bbb', language: 'fr' } ],
  _id: ObjectId('60223d51fb81142649996da6')
}
[07:44:17.747Z] INFO  EMAILTEMPLATES : [query] db.getCollection('test-template').updateMany({ _id: ObjectId('60223d51fb81142649996da6') }, { '$set': { name: 'test new name' } }, { session: undefined }); [took 44 ms]
TestTemplate {
  name: 'test new name',
  messages: [ { html: 'xxx', language: 'en' }, { html: 'yyy', language: 'fr' } ],
  _id: ObjectId('60223d51fb81142649996da6')
}

`

**Comment by B4nan** [comment]
Wait a second, why are you combining embedddables with JSON props? That is obviously wrong, either you have generic JSON property, or you have an embedded property. 

Also your PK is defined wrongly, why do you have it typed as `ObjectId` but you tell the ORM it is a `String`? 

Anyway, both of those problems won't fix the underlying issue - but that one line fix I was mentioning apparently does :]

**Comment by icex** [comment]
> Wait a second, why are you combining embedddables with JSON props? That is obviously wrong, either you have generic JSON property, or you have an embedded property.
Yeah, I was trying first with embedded before realising we can't use collections here, removing the decorator makes no difference.
With JSON as type I assumed that I can use there anything, including an array of objects.

> 
> Also your PK is defined wrongly, why do you have it typed as `ObjectId` but you tell the ORM it is a `String`?
You are correct about that PK, please ignore that. Again, this ORM is new for me.
 

> 
> Anyway, both of those problems won't fix the underlying issue - but that one line fix I was mentioning apparently does :]

Thank you kindly for your patience 👍
